multicast/src/test/kotlin/com/dropbox/flow/multicast/ChannelManagerTest.kt

```diff
@@ -15,64 +15,153 @@
+        scope.runBlockingTest {
+            val collection = async {
+                val downstream = Channel<Dispatch.Value<String>>(Channel.UNLIMITED)
+                try {
+                    manager.addDownstream(downstream)
+                    downstream.consumeAsFlow()
+                        .onEach { it.markDelivered() }
+                        .take(2)
+                        .toList()
+                        .map { it.value }
+                } finally {
+                    manager.removeDownstream(downstream)
+                }
+            }
+            try {
+                upstream.send("a")
+                upstream.send("b")
+                upstream.close()
+            } catch (e: Throwable) {
+            }
+            assertThat(collection.await()).isEqualTo(listOf("a", "b"))
+        }
+
+    @Test(expected = TestException::class)
+    fun `Given one downstream WHEN upstream errors THEN error is propagated`() =
+        scope.runBlockingTest {
+            val downstream = Channel<Dispatch.Value<String>>(Channel.UNLIMITED)
+            manager.addDownstream(downstream)
+
+            val collection = async {
+                try {
+                    downstream.consumeAsFlow()
+                        .onEach { it.markDelivered() }
+                        .take(2)
+                        .toList()
+                        .map { it.value }
+                } finally {
+                    manager.removeDownstream(downstream)
+                }
+            }
+            try {
+                upstream.close(TestException())
+            } catch (e: Throwable) {
             }
-        },
-                channelManager = it
-        )
-    }
+            collection.await()
+        }
 
     @Test
-    fun simple() = scope.runBlockingTest {
-        val collection = async {
-            val channel = Channel<DispatchValue<String>>(Channel.UNLIMITED)
+    fun `Given one downstream WHEN upstream closes THEN downstream is closed`() =
+        scope.runBlockingTest {
+            val downstream = Channel<Dispatch.Value<String>>(Channel.UNLIMITED)
+            manager.addDownstream(downstream)
+
+            val collection = async {
+                try {
+                    downstream.consumeAsFlow()
+                        .onEach { it.markDelivered() }
+                        .toList()
+                        .map { it.value }
+                } finally {
+                    manager.removeDownstream(downstream)
+                }
+            }
             try {
-                manager.send(AddChannel(channel))
-                channel.consumeAsFlow().take(2).toList()
-                    .map { it.value }
-            } finally {
-                manager.send(RemoveChannel(channel))
+                upstream.close()
+            } catch (e: Throwable) {
             }
+            delay(100) // give the upstream a chance to finish.
+            assertThat(collection.isCompleted).isTrue()
```

Comment:
couldn't we just `assertThat(collection.await()).isEmpty()` instead of the delay ?

Reply:
I don't await on downstream here to avoid the test hanging in case of a bug.